### ISSUE TITLE
problems running DM jobs at scale at SLAC

### ISSUE THREAD

**Issue by tony-johnson** [DarkEnergyScienceCollaboration/Twinkles#163]
We are currently writing the output from DM jobs to NFS space (specifically I am using /nfs/farm/g/lsst/u1/users/tonyj/). When running hundreds of jobs at once I am seeing a significant fraction (maybe 5%) failing with:

IOError: [Errno 116] Stale file handle

My guess is that this is caused by multiple jobs updating the registry.sqlite3 file. It is known that file locking is not reliable on NFS 3 mounted disks which may be causing this.

A example output file can be found here:

http://srs.slac.stanford.edu/Pipeline-II/exp/LSST-DESC/log.jsp?pi=40767767

Our normal workaround for this kind of problem is to have jobs write their output to local scratch space, and copy the output to the final location at the end of the job. I don't think DM jobs are readily amenable to this kind of fix. Another fix might be to try finding a different output area (lucene or gpfs)?

This is hopefully not fatal, since the failed jobs can be rerun (assuming that the jobs that appear to run are not in fact corrupting the registry) but it does significantly slow down processing.

**Comment by tony-johnson** [comment]
@TomGlanzman -- any suggestions about alternative disk space? I suspect we need ~200GB of space. We could move it somewhere else once the processing is complete.

**Comment by TomGlanzman** [comment]
You could try lustre.  I've added lsstsim to the group that has r/w privs in the Fermi area, but this should only be used as a scratch space.  Indeed, I am using it in the phoSim batch jobs.  Please see me for details if this is of interest.

**Comment by SimonKrughoff** [comment]
The registry.sqlite3 file is not updated after the `ingestImages.py` stage.  This must by a filesystem issue.  I don't think it should be hard to merge repos post facto, but I'm not sure there is a tool to do this yet.  I think it should be as easy as copying the repository contents to the final destination except for the `_parent` link.

**Comment by tony-johnson** [comment]
I may have been mistaken in thinking that the issue was related to the registry.sqlite file, it was just that the stack dump immiediately followed the message about opening the registry.

CameraMapper: Loading registry registry from /nfs/farm/g/lsst/u1/users/tonyj/work/0/output/_parent/registry.sqlite3
Traceback (most recent call last):
  File "/nfs/farm/g/desc/u1/Twinkles/obs_lsstSim/bin/processEimage.py", line 25, in <module>
    ProcessEimageTask.parseAndRun()

I think I will try using lustre as an experiment anyway, to see if it fixes the issue whatever it is.

**Comment by tony-johnson** [comment]
For the record, using lustre did not work at all for the ingest jobs, since it appears that (at least as configured at SLAC) no file locking is available and thus the update of the sqlite registry did not work at all.

For the other DM jobs using lustre helped somewhat, although there were still quite a lot of failures when many jobs started at once, perhaps since the DM stack software is installed on NFS.

**Comment by tony-johnson** [comment]
Stuart Marshall has set us up with some GPFS space at SLAC, which we will try using for the next DM workflow run.

**Comment by drphilmarshall** [comment]
Can we close this now we are up to speed, @tony-johnson ?

**Comment by tony-johnson** [comment]
We are still seeing job failures, although at a lower frequency than before we were using GPFS (and right now we ran out of space quota so failures are at 100%). There are a few more things to try out to improve reliability before we close this out.

**Comment by SimonKrughoff** [comment]
I think when I can get obs_LsstSim merged to master and we can upgrade to a master stack, we'll see the failures due to algorithms go down as well.

**Comment by tony-johnson** [comment]
@jchiang87 -- I still see many jobs failures in the case where many jobs all start at the same time (for examples see http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/process.jsp?min=&max=&status=NOTSUCCESS&streamIdFilter=&dateCategory=createdate&minDateDate=None&minDate=-1&maxDateDate=None&maxDate=-1&processhours=&submit=Filter&process=41531522&pstream=41535760&process=41531522). It seems to be related to nfs problems and the fact that the code installs are in nfs. Are your current code builds relocatable -- i.e. if I just do a recursive copy to copy them to AFS or GPFS and point the workflow at the copies will everything work, and read from that location instead of nfs?

**Comment by jchiang87** [comment]
I don't know if they are relocatable, but I suspect they are not.   I installed the first two (LSST_Stack, LSST_Stack_2016-02-??) using the LSST build tool (lsstsw) which builds from source.  It uses some environment variables in the setup scripts that point to the installation location, so perhaps those setup scripts just need to be updated, but there are a bunch of shared libraries, and I don't know how they were linked, so just resetting the env vars in the setup scripts may not be sufficient.  For the LSST_Stack_2016-04-12 install, I used the conda distribution, which may or may not be relocatable.  I simply do not know.   However, re-installing the conda distr. in a new location should be relatively painless.

**Comment by SimonKrughoff** [comment]
@jchiang87 @tony-johnson I just installed a new stack at SLAC.  It's located at `/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw`.

The most recent build with the DM-4302 merge is b2016.  I didn't make a nice setup like Jim had, but it is basically:
```
$> scl enable devtoolset-3 bash
$> source bin/setup.sh
$> setup lsst_apps -t b2016
```

**Comment by tony-johnson** [comment]
I would like to try this out with the workflow, but currently the workflow assumes the existence of Jim's setup script. I could probably create this myself based on previous examples, but currently I don't have write access to the installation dir. @SimonKrughoff -- could you:

chmod -R  g+w /nfs/slac/kipac/fs1/g/desc/DMStack

(or add something equivalent to Jim's setup.sh).

**Comment by jchiang87** [comment]
@tony-johnson I'm in the process of running the scripts on a small dataset to make sure everything works.  Note that the ingestImages.py script has been renamed to ingestSimImages.py and now every script needs to have an --output specified (see https://github.com/DarkEnergyScienceCollaboration/Twinkles/commit/e3b29c42ef19bf9b1723ed7c9cb261e108f746ad ).   I'll put a set up script somewhere I have write access shortly.

**Comment by jchiang87** [comment]
Ok, the setup command is
```
/nfs/slac/kipac/fs1/g/desc/run_setup
```

**Comment by jchiang87** [comment]
I ran the cookbook scripts on the first ten visits of the `Twinkles-phoSimII` data.  There were failures in one of the `procesEimage.py` runs, in `mergeCoaddDetections.py`, `measureCoaddSources.py`, and `forcedPhotCcd.py`.   Presumably the failure in `mergeCoaddDetections.py` propagated into the latter two. 
```
processEimage.py image_repo/ --id visit=2226 --output output_repo --doraise --clobber-config
```

```
processEimage.calibrate: Processing {'filter': 'u', 'sensor': '1,1', 'visit': 2226, 'raft': '2,2'}
Traceback (most recent call last):
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/obs_lsstSim/2016_01.0-15-g69e1fce/bin/processEimage.py", line 25, in <module>
    ProcessEimageTask.parseAndRun()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 450, in parseAndRun
    resultList = taskRunner.run(parsedCmd)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 199, in run
    resultList = mapFunc(self, targetList)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 324, in __call__
    result = task.run(dataRef, **kwargs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/timer.py", line 118, in wrapper
    res = func(self, *args, **keyArgs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/processCcd.py", line 171, in run
    icSourceCat = charRes.sourceCat,
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/timer.py", line 118, in wrapper
    res = func(self, *args, **keyArgs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/calibrate.py", line 321, in run
    icSourceCat = icSourceCat,
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/calibrate.py", line 362, in calibrate
    background = background,
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/timer.py", line 118, in wrapper
    res = func(self, *args, **keyArgs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/detectAndMeasure.py", line 192, in run
    detRes = self.detection.run(table=table, exposure=exposure, doSmooth=True)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/timer.py", line 118, in wrapper
    res = func(self, *args, **keyArgs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_algorithms/2016_01.0-9-g3ec4b9e+1/python/lsst/meas/algorithms/detection.py", line 345, in run
    clearMask=clearMask)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/timer.py", line 118, in wrapper
    res = func(self, *args, **keyArgs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_algorithms/2016_01.0-9-g3ec4b9e+1/python/lsst/meas/algorithms/detection.py", line 424, in detectFootprints
    kWidth = (int(sigma * 7 + 0.5) // 2) * 2 + 1 # make sure it is odd
ValueError: cannot convert float NaN to integer
```
----
```
mergeCoaddDetections.py output_repo/ --id filter=u^g^r^i^z^y patch=0,0 tract=0 --output output_repo --doraise --clobber-config
```

```
Traceback (most recent call last):
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/bin/mergeCoaddDetections.py", line 3, in <module>
    MergeDetectionsTask.parseAndRun()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 450, in parseAndRun
    resultList = taskRunner.run(parsedCmd)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 199, in run
    resultList = mapFunc(self, targetList)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 324, in __call__
    result = task.run(dataRef, **kwargs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/multiBand.py", line 446, in run
    mergedCatalog = self.mergeCatalogs(catalogs, patchRefList[0])
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/multiBand.py", line 708, in mergeCatalogs
    self.schema, self.makeIdFactory(patchRef),
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/multiBand.py", line 88, in makeIdFactory
    return afwTable.IdFactory.makeSource(expId, 64 - expBits)
  File "<string>", line 5, in __rsub__
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/readProxy.py", line 41, in __getattribute__
    subject = oga(self, '__subject__')
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/readProxy.py", line 136, in __subject__
    set_cache(self, get_callback(self)())
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 363, in <lambda>
    callback = lambda: bypassFunc(datasetType, pythonType, location, dataId)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/obs_lsstSim/2016_01.0-15-g69e1fce/python/lsst/obs/lsstSim/lsstSimMapper.py", line 225, in bypass_deepMergedCoaddId_bits
    return 64 - self._nbit_id
AttributeError: 'LsstSimMapper' object has no attribute '_nbit_id'
```
----
```
measureCoaddSources.py output_repo/ --id filter=u^g^r^i^z^y patch=0,0 tract=0 --config measurement.doApplyApCorr='yes' --output output_repo --doraise --clobber-config
```

```
Traceback (most recent call last):
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/bin/measureCoaddSources.py", line 3, in <module>
    MeasureMergedCoaddSourcesTask.parseAndRun()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 450, in parseAndRun
    resultList = taskRunner.run(parsedCmd)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 199, in run
    resultList = mapFunc(self, targetList)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 324, in __call__
    result = task.run(dataRef, **kwargs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/multiBand.py", line 1032, in run
    sources = self.readSources(patchRef)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_tasks/2016_01.0-34-g70cb220+1/python/lsst/pipe/tasks/multiBand.py", line 1072, in readSources
    merged = dataRef.get(self.config.coaddName + "Coadd_mergeDet", immediate=True)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butlerSubset.py", line 192, in get
    return self.butlerSubset.butler.get(datasetType, self.dataId, **rest)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 370, in get
    return callback()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 365, in <lambda>
    callback = lambda: self._read(location)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 455, in _read
    results = location.repository.read(location)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/repository.py", line 346, in read
    return self._access.read(butlerLocation)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/access.py", line 125, in read
    return self.storage.read(butlerLocation=butlerLocation)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/posixStorage.py", line 251, in read
    raise RuntimeError, "No such FITS catalog file: " + logLoc.locString()
RuntimeError: No such FITS catalog file: /nfs/farm/g/desc/u1/users/jchiang/desc_projects/twinkles/DM-4302_test/Run1_tests/output_repo/deepCoadd-results/merged/0/0,0/mergeDet-0-0,0.fits
```
----
```
forcedPhotCcd.py output_repo/ --id tract=0 visit=2226^2942^3665^4571^220^200^185870^230^250^276 sensor=1,1 raft=2,2 --config measurement.doApplyApCorr='yes' --output output_repo --doraise --clobber-config
```

```
Traceback (most recent call last):
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_base/2016_01.0-14-g095ee0a+5/bin/forcedPhotCcd.py", line 24, in <module>
    ForcedPhotCcdTask.parseAndRun()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 450, in parseAndRun
    resultList = taskRunner.run(parsedCmd)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 199, in run
    resultList = mapFunc(self, targetList)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/pipe_base/2016_01.0-7-gee41fc9+1/python/lsst/pipe/base/cmdLineTask.py", line 324, in __call__
    result = task.run(dataRef, **kwargs)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_base/2016_01.0-14-g095ee0a+5/python/lsst/meas/base/forcedPhotImage.py", line 121, in run
    exposure = self.getExposure(dataRef)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_base/2016_01.0-14-g095ee0a+5/python/lsst/meas/base/forcedPhotCcd.py", line 235, in getExposure
    exposure = ProcessImageForcedTask.getExposure(self, dataRef)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/meas_base/2016_01.0-14-g095ee0a+5/python/lsst/meas/base/forcedPhotImage.py", line 181, in getExposure
    return dataRef.get(self.dataPrefix + "calexp", immediate=True)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butlerSubset.py", line 192, in get
    return self.butlerSubset.butler.get(datasetType, self.dataId, **rest)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 370, in get
    return callback()
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 368, in <lambda>
    callback = lambda: location.mapper.standardize(datasetType, innerCallback(), dataId)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 365, in <lambda>
    callback = lambda: self._read(location)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/butler.py", line 455, in _read
    results = location.repository.read(location)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/repository.py", line 346, in read
    return self._access.read(butlerLocation)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/access.py", line 125, in read
    return self.storage.read(butlerLocation=butlerLocation)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/posixStorage.py", line 265, in read
    butlerLocation.getCppType(), storageList, additionalData)
  File "/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/Linux64/daf_persistence/2016_01.0-8-geb625e2+3/python/lsst/daf/persistence/persistenceLib.py", line 1430, in unsafeRetrieve
    return _persistenceLib.Persistence_unsafeRetrieve(self, *args)
lsst.afw.fits.fitsLib.FitsError: 
  File "src/fits.cc", line 1067, in lsst::afw::fits::Fits::Fits(const string&, const string&, int)
    cfitsio error: could not open the named file (104) : Opening file '/nfs/farm/g/desc/u1/users/jchiang/desc_projects/twinkles/DM-4302_test/Run1_tests/output_repo/calexp/v2226-fu/R22/S11.fits' with mode 'r' {0}
lsst::afw::fits::FitsError: 'cfitsio error: could not open the named file (104) : Opening file '/nfs/farm/g/desc/u1/users/jchiang/desc_projects/twinkles/DM-4302_test/Run1_tests/output_repo/calexp/v2226-fu/R22/S11.fits' with mode 'r''
```

**Comment by SimonKrughoff** [comment]
The nbit bug is fixed on master.  I'm building a new version of the stack now (it will be b2017).  I am working with @wmwv to put some continuous integration in for processing of simulated images to try to avoid bugs like this.  That should fix the error you are seeing in mergeCoaddDetections.

To be clear, there was one failure in processEimage and the rest succeeded?  I'll try to reproduce the error.

**Comment by jchiang87** [comment]
yes, 9 of 10 succeeded for processEimage.

**Comment by jchiang87** [comment]
BTW, the mergeCoaddDetections.py failure didn't seem to return a non-zero return code even though I ran with the --doraise option.  The other three failures had non-zero return codes.  FWIW, I using my [Level2_Pipeline](https://github.com/jchiang87/Level2_Pipeline) to exercise the cookbook scripts.

**Comment by tony-johnson** [comment]
I also tried to run using the new code build, but I noticed that the serialIngest stage has slowed down enormously. Previously it took about 5000 CPU second to process 900 odd visits, now it run out of CPU time after 7160 seconds. Is this expected? My understanding was that it was basically creating symlinks, so I am not sure why it was slow to begin with?

**Comment by SimonKrughoff** [comment]
@tony-johnson The ingest should not have changed at all.  It is creating symlinks, but it does need to read the header from the fits file.  How many were ingested in 7000 sec?

I just tested this on my laptop for 30 visits and it took 22 sec.

**Comment by SimonKrughoff** [comment]
@tony-johnson @jchiang87 FYI, I'm doing some debugging on the stack, so don't run any large jobs for the moment.  I'm inserting calls to pdb so I can look at values in the processing.

**Comment by tony-johnson** [comment]
Strange, it processed 811 visits in that time. I will investigate and 
see if I can find any hints as to why it is running so slowly.

**Comment by SimonKrughoff** [comment]
O.K.  I added a config override to obs_lsstSim that seems to help.  I'm rebuilding the stack because something got all confused.  I'll ping this thread when it has built.

**Comment by tony-johnson** [comment]
A stupid question, where do I look to find what ingestSimImages actually runs? I found 

./build/obs_lsstSim/bin.src/ingestSimImages.py
#!/usr/bin/env python
from lsst.obs.lsstSim.ingest import SimIngestTask
SimIngestTask.parseAndRun()

but where do I look to find the definition of SimIngestTask?

**Comment by SimonKrughoff** [comment]
You can look in $OBS_LSSTSIM_DIR/python/lsst/obs/lsstSim/ingest.py.

**Comment by SimonKrughoff** [comment]
@jchiang87 @tony-johnson I installed the new version of the stack in the same location.  The tag to use is b2016.

**Comment by SimonKrughoff** [comment]
I ran processEimage.py on the first 40 visits in the dataset and they all worked, so I'm hoping we are on the right track now.

**Comment by jchiang87** [comment]
@SimonKrughoff Do you mean b2017 (not b2016)?   BTW, when I try to see which tags area available for obs_lsstSim, I get this:
```
bash-4.1$ eups list obs_lsstSim
eups list: [Errno 13] Permission denied: '/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw/build/lsstsw/stack/ups_db/Linux64.pickleDB1_3_0'
bash-4.1$ 
```

**Comment by SimonKrughoff** [comment]
No I meant b2016.  The last build overwrote one of my previous tags for some reason.

Oh man.  Now I see the problem.  I accidentally started another stack while the first was still set up.  Let me rebuild the stack in the right place.  Sorry about that.

**Comment by SimonKrughoff** [comment]
O.K.  I deleted the other offending stack.  There is only one now (I hope) in: /nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw

If you start a new shell and source the `bin/setup.sh` script in there, you should only see a single b2016 tag.  That's the one to use.

**Comment by jchiang87** [comment]
ok, thanks.  @tony-johnson I've updated the /nfs/slac/kipac/fs1g/desc/setup.sh script to use that location.

**Comment by tony-johnson** [comment]
I have create two new streams using @SimonKrughoff latest build:

Stream|Comment|URL
-------|-----------|---
403|Using same 984 visits as before|http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/si.jsp?stream=41541421
500|Using all 1124 visits which completed phosim|http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/si.jsp?stream=41541457

Unfortunately in both cases (the same) 4 visits failed in processEimage

Visit|Error|Log File
-----|-----|-----------
681808 | ZeroDivisionError: float division by zero | http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/log.jsp?pi=41545743
1417225 | ZeroDivisionError: float division by zero | http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/log.jsp?pi=41545746
1957025 | ValueError: cannot convert float NaN to integer | http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/log.jsp?pi=41545751
2166623 | ValueError: cannot convert float NaN to integer | http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/log.jsp?pi=41545749

For now I will create a new stream (501) which skips these visits. Incidentally I am still seeing very slow ingest jobs. I have verified that this is not being caused by nfs (I copied all the phosim output to gpfs but saw little difference in performance). I have further verified that I get much better throughput when running interactively than in batch. I think it may be an anomaly in the way the SLAC batch system schedules IO intensive jobs. For now it is not a show stopper (just makes the show reaaalllyyy slow).

**Comment by SimonKrughoff** [comment]
@tony-johnson thanks for this.  For what it's worth, you don't have to do the ingest step every time.  I understand that it's nice to do the whole process, but we could just use the same input repo for all the runs.

**Comment by tony-johnson** [comment]
We are continuing to see problems with very long elapsed times for batch jobs (right now the coadd jobs for stream 501). This has been reported to SLAC unix-admin (as [INC0092281](https://slacprod.service-now.com/nav_to.do?uri=incident.do%3Fsys_id=b10a60d56f031240fd565d412e3ee4bc%26sysparm_stack=incident_list.do%3Fsysparm_query=active=true))

**Comment by tony-johnson** [comment]
Run 501 is now complete (at last). This run processed 1120 visits (all that were successfully processed through phosim) minus 4 which failed in processEimage (documented above), using the DMStack which Simon installed here:

/nfs/slac/kipac/fs1/g/desc/DMStack/lsstsw

The output can be found at SLAC here:

/nfs/slac/kipac/fs1/g/desc/Twinkles/501

The workflow (useful for looking at log files etc) can be found here:

http://srs.slac.stanford.edu/Pipeline-II/exp/SRS/si.jsp?stream=41545777

Hopefully Jim and Simon and others can look at the output and decide if we can declare this to be Run 1.1, or if more work is still needed. 

I did get some response from SLAC unix-admin concerning the slowness of the batch jobs, with some suggestion that the slowness may be caused by MPI jobs getting unfair share of CPU time on the bullet cluster. I am not yet convinced this is the cause, but the batch farm will be reconfigured next Monday to deal with this issue, so hopefully future runs be much quicker.

**Comment by drphilmarshall** [comment]
Hooray! Well done Tony :-)

I guess the immediate validation tests are: look at the color composite of the stacked image, check objects are in the right place? Checking light curves can be done once the ForcedSources are easily queryable. I guess @wmwv 's `drp_validate` script can be run - is that not part of our pipeline yet? That task was issued in #221.  All this validation feels like a separate issue, though, part of #167 I'd be inclined to call this Run 1.1, ship it to NERSC, and start the analysis work. Sound good?

**Comment by jchiang87** [comment]
Awesome!  Yes, let's ship it to NERSC, and once it's there, I'll ingest it into the database.

**Comment by SimonKrughoff** [comment]
I'll make a color composite tonight and send it around.

**Comment by jchiang87** [comment]
This isn't a color composite, rather it's a set of  20x20 arcsec postage stamps centered on an object picked at random (objectId=40390) with all of the objects from the merged catalog in the field overlaid on the image for each band.  Looks good to me.

![40390_ugrizy_stamps](https://cloud.githubusercontent.com/assets/1994473/14873494/1bf3f42e-0cac-11e6-8c06-d25afb3d0fd1.png)

**Comment by SimonKrughoff** [comment]
Really awesome @jchiang87 !

**Comment by tony-johnson** [comment]
Phil, can you remind me where the group space is at NERSC which we discussed on Friday - once I have that I will copy the data there.

**Comment by SimonKrughoff** [comment]
Color images with full gri, subsection of ury, and another thumbnail from giy.  Let me know if you want the other two full images.  I just couldn't attach them to this comment.
![image](https://cloud.githubusercontent.com/assets/945715/14874645/529e031a-0cb7-11e6-92a7-a000c8514d11.png)
![image](https://cloud.githubusercontent.com/assets/945715/14874687/a333652c-0cb7-11e6-9859-7dbd3f4c4476.png)
![image](https://cloud.githubusercontent.com/assets/945715/14874693/b19ed790-0cb7-11e6-9cb3-6df8d6db64d0.png)

**Comment by drphilmarshall** [comment]
Sweet! :-) Looks like the registration and object deblending is working really nicely, check out those 1-2" source separations in Jim's cutouts! 

@jbkalmbach @cdfassnacht we have some lens validation to do - some systems look right, others do not (but I could be getting confused by the blue stars). Here's a nice zoom of two lenses that look good:

![image](https://cloud.githubusercontent.com/assets/710903/14875163/7e2daef4-0cbc-11e6-88a3-748bcd89b3d8.png)

So perhaps they are all good? 

Anyway, great work everyone! @tony-johnson I think you should close this out :-)  The NERSC project  area is `/project/projectdirs/lsst/`, according to https://confluence.slac.stanford.edu/display/LSSTDESC/Getting+a+NERSC+Computing+Account Thanks!

**Comment by SimonKrughoff** [comment]
I like this one.  @drphilmarshall do you have any idea what's going on here?
![image](https://cloud.githubusercontent.com/assets/945715/14875301/340cf6c0-0cbe-11e6-8e97-72733ff49e65.png)

**Comment by drphilmarshall** [comment]
No! I'd really like to see the CatSim stars and galaxy positions overlaid,
not to mention the OM10 object positions!

**Comment by tony-johnson** [comment]
The run 501 data has now been copied to NERSC (cori) in:

/global/homes/t/tony_j/Twinkles/Run1.1/
